ql/instruments/forwardrateagreement.cpp

```diff
@@ -57,12 +58,67 @@ namespace QuantLib {
                                                Rate strikeForwardRate,
                                                Real notionalAmount,
                                                const ext::shared_ptr<IborIndex>& index,
+                                               Handle<YieldTermStructure> discountCurve,
+                                               bool useIndexedCoupon)
+    : fraType_(type), notionalAmount_(notionalAmount), index_(index),
+      useIndexedCoupon_(useIndexedCoupon), dayCounter_(index->dayCounter()),
+      calendar_(index->fixingCalendar()), businessDayConvention_(index->businessDayConvention()),
+      valueDate_(valueDate), maturityDate_(index->maturityDate(valueDate)),
+      discountCurve_(std::move(discountCurve)) {
+
+        maturityDate_ = calendar_.adjust(maturityDate_, businessDayConvention_);
+
+        registerWith(Settings::instance().evaluationDate());
+        registerWith(discountCurve_);
+
+        QL_REQUIRE(notionalAmount > 0.0, "notionalAmount must be positive");
+        QL_REQUIRE(valueDate_ < maturityDate_, "valueDate must be earlier than maturityDate");
+
+        strikeForwardRate_ = InterestRate(strikeForwardRate,
+                                          index->dayCounter(),
+                                          Simple, Once);
+        registerWith(index_);
+    }                           
+
+    ForwardRateAgreement::ForwardRateAgreement(const Date& valueDate,
+                                               const Date& maturityDate,
+                                               Position::Type type,
+                                               Rate strikeForwardRate,
+                                               Real notionalAmount,
+                                               //Handle<YieldTermStructure> discountCurve,
+                                               Handle<YieldTermStructure> forecastCurve,
+                                               Natural fixingDays,
+                                               BusinessDayConvention businessDayConvention,
                                                Handle<YieldTermStructure> discountCurve)
-    : ForwardRateAgreement(valueDate, index->maturityDate(valueDate), type, strikeForwardRate,
-                           notionalAmount, index, std::move(discountCurve), true) {}
+    : fraType_(type), notionalAmount_(notionalAmount),
+      useIndexedCoupon_(false), dayCounter_(discountCurve->dayCounter()),
```

Comment:
Here's the problem — the discount curve might be empty.  Since we were taking the day counter from the index, I would probably add it as another parameter to the constructor, like the fixing days and the convention.  The same goes for the calendar, which was also taken from the index.